### DISCUSSION TITLE
Add a universal `j.u.l.LogManager` implementation

### DISCUSSION THREAD

**Discussion by ppkarwasz** [qos-ch/slf4j#429] [Ideas]
As most developers are painfully aware redirecting `java.util.logging` to a decent logging framework is hard and:

1. it either requires to modify the `java.util.logging.manager` property very early in the lifecycle of an application, which is highly impractical,
2. or it requires the usage of a redirecting [`j.u.l.Handler`](https://docs.oracle.com/en/java/javase/21/docs/api/java.logging/java/util/logging/Handler.html) like [`SLF4JBridgeHandler`](https://www.slf4j.org/apidocs/org/slf4j/bridge/SLF4JBridgeHandler.html), which is less efficient. Log events traverse the entire JUL logging pipeline before being injected into SLF4J.

The best solution would be to force `j.u.l.LogManager` to use `ServiceLoader`, but I don't believe [JDK-8262741](https://bugs.openjdk.org/browse/JDK-8262741) will be solved by OpenJDK any time soon.

What do you think about creating an `org.slf4j.jul.LogManager` implementation, whose only purpose would be to load some sort of `LogManagerAdapter` implementation via `ServiceLoader`? If such a solution were to be logging API-agnostic, many application runtime could hardcode the value of `java.util.logging.manager` in their shell scripts or main classes, while leaving the user the freedom to choose the `JUL-to-X` adapter of their choice.

**Comment by ppkarwasz** [comment]
I have provided an initial mock-up of such a universal `j.u.l.LogManager` implementation in [ppkarwasz/slf4j/jul-manager](https://github.com/ppkarwasz/slf4j/tree/feature/jul-manager/jul-manager).

**Comment by ceki** [comment]
@ppkarwasz Thank you for opening this discussion. I have not dabbled in `LogManager` in a long time. Could you please provide an example how the application runtime would hard code the value of java.util.logging.manager in their shell scripts or main classes?

**Comment by ppkarwasz** [comment]
The best example of setting  the `j.u.l.LogManager` in a shell script is Tomcat's `catalina.sh`: it uses a [`LOGGING_MANAGER` shell variable](https://github.com/apache/tomcat/blob/2681cfc34a945df92e2cd596134d0363f2424249/bin/catalina.sh#L102-L105) that sets up the `java.util.logger.manager`.

On the other hand Wildfly sets the `java.util.logging.manager` property in its code, e.g. [`org.jboss.modules.Main`](https://github.com/jboss-modules/jboss-modules/blob/bcbc19ab473bf167a1d7886e9ec9af0e73b6a3d0/src/main/java/org/jboss/modules/Main.java#L437-L445).

A rapid [GitHub search for `o.a.l.l.j.LogManager`](https://github.com/search?q=org.apache.logging.log4j.jul.LogManager+language%3Ajava&type=code) shows that many applications hardcode `o.a.l.l.j.LogManager` in their main class.

I would pretty much prefer if these 3 examples used the **same** implementation of `j.u.l.LogManager`, which would delegate calls of `LogManager.getLogger()` to the appropriate implementation retrieved from `ServiceLoader` (Tomcat's implementation of `LogManager`, JBoss Logging or Log4j API in these cases).
